---
title: "BIS15L Final Project- PharmGKB relationships"
output: html_document
date: "2025-03-11"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Project Goals
1. How does Gene ABCB1 associate with different diseases?
2. How strong is the evidence for the association between Variant X and different disease?  

##Resources  
- [Warfarin data](https://www.pharmgkb.org/downloads)  

## Install the packages
```{r}
if(!requireNamespace("BiocManager")){
  install.packages("BiocManager")
}
BiocManager::install("biomaRt")
```

```{r}
install.packages("visNetwork")
```


## Load the libraries
```{r}
library(biomaRt)
library(tidyverse)
library(janitor)
library(dplyr)
library(shiny)
library(shinydashboard)
require(janitor)
library(visNetwork)
```

## Data
```{r}
relations <- read_tsv("data/relationships.tsv") %>% clean_names()
```

```{r}
names(relations)
glimpse(relations)
```

##Gene vs Disease
##Gene ABCB1 Analysis

#Find the most common gene as entity 1.
```{r}
relations %>% 
  filter(entity1_type == "Gene") %>% 
  count(entity1_name, sort = T)

# The most common gene appears to be ABCB1!
```

#Filter the data set to have only the data we want.
```{r}
filtered_relations <- relations %>% 
  select(-pk, -pd) %>% 
  filter(entity1_name == "ABCB1" & entity2_type == "Disease" & association == "associated")
filtered_relations
```

#Count the number of diseases
```{r}
disease_count <- filtered_relations %>% 
  count(entity2_name, sort = T)
```

#Categorize the diseases
```{r}
disease_count <- disease_count %>%
  mutate(Category = case_when(
    grepl("cancer|carcinoma|lymphoma|tumor", entity2_name, ignore.case = TRUE) ~ "Cancer",
    grepl("diabetes", entity2_name, ignore.case = TRUE) ~ "Metabolic",
    grepl("hypertension|cardiac|coronary|arrhythmia", entity2_name, ignore.case = TRUE) ~ "Cardiovascular",
    grepl("autism|attention deficit|nervous system", entity2_name, ignore.case = TRUE) ~ "Neurological",
    TRUE ~ "Other"
  ))

# Count number of diseases per category
category_counts <- disease_count %>%
  count(Category, sort = TRUE)

# Plot disease categories
ggplot(category_counts, aes(x = reorder(Category, n), y = n)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  labs(title = "Disease Categories Associated with ABCB1",
       x = "Disease Category",
       y = "Number of Associations") +
  scale_fill_brewer(palette = "Accent")+
  theme_light()

# ABCB1 is associated with a wide variety of diseases that donâ€™t fit into common categories like cancer, cardiovascular, neurological, or metabolic diseases.

# ABCB1 is known for its role in drug transport and multidrug resistance, meaning it might be linked to diverse conditions rather than a specific disease class.

# to show evidence type, filter out clinical or variant, focus with cancer
```

#Analyze cancer
```{r}
# Filter cancer-related diseases and exclude "clinical" or "variant" evidence types
cancer_evidence <- filtered_relations %>%
  mutate(Category = case_when(
    grepl("cancer|carcinoma|lymphoma|tumor", entity2_name, ignore.case = TRUE) ~ "Cancer",
    grepl("diabetes", entity2_name, ignore.case = TRUE) ~ "Metabolic",
    grepl("hypertension|cardiac|coronary|arrhythmia", entity2_name, ignore.case = TRUE) ~ "Cardiovascular",
    grepl("autism|attention deficit|nervous system", entity2_name, ignore.case = TRUE) ~ "Neurological",
    TRUE ~ "Other"
  )) %>%
  filter(Category == "Cancer")
```


#Get the Table for cancer associated with its evidence
```{r}
cancer_disease_evidence <- cancer_evidence %>%
  select(entity2_name, evidence) %>%
  arrange(entity2_name)

# View the first few rows
head(cancer_disease_evidence)

# Count number of times each (disease, evidence) pair appears
cancer_disease_evidence_counts <- cancer_evidence %>%
  count(entity2_name, evidence, sort = TRUE)

# Check the counts
head(cancer_disease_evidence_counts)
```

#Plotting the result
```{r}
ggplot(cancer_disease_evidence_counts, aes(x = evidence, y = n, fill = evidence)) +
  geom_bar(stat = "identity") +
  labs(title = "Cancer-Related Diseases Vs. Evidence Types",
       x = "Disease",
       y = "Number of Associations",
       fill = "Evidence Type") +
  scale_fill_brewer(palette = "Accent")+
  theme_light()

# Basically different types of Lymphoma are clinical+variant associated, and carcinoma is variant associated.
```



#PI Chart
```{r}
ggplot(cancer_disease_evidence_counts, aes(x = "Disease", y = n, fill = evidence)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar(theta = "y") +
  labs(title = "Distribution of Evidence Types for Cancer-Related Diseases",
       fill = "Evidence Type") +
  scale_fill_brewer(palette = "Accent")+
  theme_void()
```


#Conclusion
```{r}
# Basically Lymphoma is more well supported by clinical and variant evidence, and is linked to the gene of interest. It implies that there is a strong genetic link between Lymphoma and ABCB1.
```



##Variant vs Disease

#Dataframe with only variant genes and diseases
```{r}
variant_disease <- relations %>% 
  filter(entity1_type == "Variant" & entity2_type == "Disease")
```

#Percentage of association between variant genes and diseases
```{r}
pa_variant_disease <- variant_disease %>% 
  group_by(association) %>%
  summarise(count = n(), .groups = 'drop') %>%
  mutate(percentage = (count / sum(count)) * 100) %>% 
  arrange(-count)
pa_variant_disease 
```

#Graph of percentage of association between variant genes and diseases
```{r}
ggplot(pa_variant_disease, aes(x = "", y = percentage, fill = association)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar("y", start = 0) +
  labs(title = "Association Types between Variant Genes and Diseases",
       fill= "Type of Association") +
  scale_fill_brewer(palette = "Accent")+
  theme_void()+
  geom_text(aes(label = paste0(round(percentage, 1), "%")),  # Adding percentage labels
            position = position_stack(vjust = 0.5),             # Centers the labels
            color = "white", size = 4)
```


#Which gene is most associated with having a relationship with a disease?
```{r}
variant_disease %>% 
  group_by(entity1_name) %>%
  filter(association == "associated") %>% 
  summarise(count = n(), .groups = 'drop') %>%
  mutate(percentage = (count / sum(count)) * 100) %>% 
  arrange(-percentage)
```
The presence of the variant rs1045642 gene is most associated with having a disease out of all the other variant genes.

#Graph of genes most associated with having a relationship with a disease
```{r}
variant_disease %>% 
  group_by(entity1_name, association) %>%
  summarise(count = n(), .groups = 'drop') %>%
  mutate(percentage = (count/sum(count)) * 100) %>%
  arrange(-count) %>%
  slice_max(order_by = count, n = 10) %>%
  ggplot(aes(x = reorder(entity1_name, percentage), y = percentage, fill = association)) +
  geom_bar(stat = "identity", position = "dodge") +
  coord_flip() +
  labs(title = "Association Status by Variant Gene",
       x = "Variant Gene",
       y = "Percentage (%)",
       fill = "Association Type") +
  scale_fill_brewer(palette = "Accent")+
  theme_light()
```

#Which diseases is the rs1045642 gene associated with?
```{r}
variant_disease %>% 
  select(entity1_name, entity2_name, association) %>% 
  filter(entity1_name == "rs1045642" & association == "associated")
```

#What diseases are most associated with variant genes?
```{r}
variant_disease %>% 
  group_by(entity2_name) %>%
  filter(association == "associated") %>% 
  summarise(count = n(), .groups = 'drop') %>%
  mutate(percentage = (count / sum(count)) * 100) %>% 
  arrange(-percentage)
```

#Graph of most associated diseases 
```{r}
variant_disease %>% 
  group_by(entity2_name, association) %>%
  summarise(count = n(), .groups = 'drop') %>%
  mutate(percentage = (count/sum(count)) * 100) %>%
  arrange(-count) %>%
  slice_max(order_by = count, n = 10) %>%
  ggplot(aes(x = reorder(entity2_name, percentage), y = percentage, fill = association)) +
  geom_bar(stat = "identity", position = "dodge") +
  coord_flip() +
  labs(title = "Association Status by Variant Gene",
       x = "Disease",
       y = "Percentage (%)",
       fill = "Association Type") +
  scale_fill_brewer(palette = "Accent")+
  theme_light()
```

#Categorize diseases: MODIFY- innacurate
```{r}
disease_count_v <- variant_disease %>%
  mutate(Category = case_when(
    grepl("cancer|carcinoma|lymphoma|tumor", entity2_name, ignore.case = TRUE) ~ "Cancer",
    grepl("diabetes", entity2_name, ignore.case = TRUE) ~ "Metabolic",
    grepl("hypertension|cardiac|coronary|arrhythmia", entity2_name, ignore.case = TRUE) ~ "Cardiovascular",
    grepl("autism|attention deficit|nervous system", entity2_name, ignore.case = TRUE) ~ "Neurological",
    TRUE ~ "Other"
  ))

category_counts <- disease_count_v %>%
  count(Category, sort = TRUE)

ggplot(category_counts, aes(x = reorder(Category, n), y = n)) +
  geom_bar(stat = "identity", fill = "steelblue") + ###CHANGE COLOR
  coord_flip() +
  labs(title = "Disease Categories Associated with Variant Genes",
       x = "Disease Category",
       y = "Number of Associations") +
  scale_fill_brewer(palette = "Accent")+
  theme_light()
```

#Comparing amount of evidence for associated vs. not associated interactions
```{r}
ev_assoc <- variant_disease %>% 
  separate_rows(pmi_ds, sep = ";") %>% 
  group_by(association) %>%
  summarize(pmid_count = n())

head(ev_assoc)
```

#Analysis of clinical vs. variant annotation
```{r}
annotation_count_vd <- variant_disease %>% 
  filter(association == "associated") %>% 
  group_by(evidence) %>%
  summarise(count = n(), .groups = 'drop') %>%
  mutate(percentage_e = (count / sum(count)) * 100) %>% 
  arrange(-count)
annotation_count_vd
```

#Cancer
```{r}
# Filter cancer-related diseases and exclude "clinical" or "variant" evidence types
cancer_evidence_v <- variant_disease %>%
  mutate(Category = case_when(
    grepl("cancer|carcinoma|lymphoma|tumor", entity2_name, ignore.case = TRUE) ~ "Cancer",
    grepl("diabetes", entity2_name, ignore.case = TRUE) ~ "Metabolic",
    grepl("hypertension|cardiac|coronary|arrhythmia", entity2_name, ignore.case = TRUE) ~ "Cardiovascular",
    grepl("autism|attention deficit|nervous system", entity2_name, ignore.case = TRUE) ~ "Neurological",
    TRUE ~ "Other"
  )) %>%
  filter(Category == "Cancer")
cancer_evidence_v
```

```{r}
cancer_disease_evidence_v <- cancer_evidence_v %>%
  select(entity2_name, evidence) %>%
  arrange(entity2_name)

# View the first few rows
head(cancer_disease_evidence_v)

# Count number of times each (disease, evidence) pair appears
cancer_disease_evidence_counts_v <- cancer_evidence_v %>%
  count(entity2_name, evidence, sort = TRUE)

# Check the counts
head(cancer_disease_evidence_counts_v)
```

#Graph of cancers vs. evidence associated in filtered dataset (variant genes and disease)
```{r}
ggplot(cancer_disease_evidence_counts_v, aes(x = reorder(entity2_name, n), y = n, fill = evidence)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  labs(title = "Cancer-Related Diseases Vs. Evidence Types",
       x = "Disease",
       y = "Number of Associations",
       fill = "Evidence Type") +
  scale_fill_brewer(palette = "Accent")+
  theme_light()

# Basically different types of Lymphoma are clinical+variant associated, and carcinoma is variant associated.
```

#Graph of evidence types and count in filtered dataset (variant genes and disease)
```{r}
ggplot(annotation_count_vd, aes(x = evidence, y = count, fill = evidence)) +
  geom_bar(stat="identity") +
  scale_y_log10()+
  labs(title = "Evidence Types",
       x = "Evidence Type",
       y = "Count")+
    theme(axis.text.x = element_text(angle = 60, hjust = 1))+
    scale_fill_brewer(palette = "Accent")+
  theme(legend.position = "none")

#Gives alternate more of quantitative result: how many of these associations are each type of evidence
```



##OVERALL CONCLUSION

#Association with variant
```{r}
# Connect to Ensembl
ensembl <- useMart(biomart = "ENSEMBL_MART_SNP", dataset = "hsapiens_snp")

variant_info <- getBM(
  attributes = c("refsnp_id", "chr_name", "chrom_start", "ensembl_gene_stable_id"),
  filters = "snp_filter",
  values = "rs1045642",  # Replace with your variant ID
  mart = ensembl
)

#print(variant_info)

ensembl_genes <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl")

# Get the HGNC symbol for the Ensembl gene ID
gene_info <- getBM(
  attributes = c("ensembl_gene_id", "hgnc_symbol"),
  filters = "ensembl_gene_id",
  values = "ENSG00000085563",  # Your gene ID
  mart = ensembl_genes
)

print(gene_info)
```

#SHINY APP


#Graph of cancers vs. evidence associated in filtered dataset (variant genes and disease)
```{r}
cancer_disease_evidence_counts_v_rs <- cancer_evidence_v %>% 
  filter(entity1_name == "rs1045642") %>% 
  count(entity2_name, evidence, sort = TRUE)

ggplot(cancer_disease_evidence_counts_v_rs, aes(x = evidence, y = n, fill = evidence)) +
  geom_bar(stat = "identity") +
  labs(title = "Cancer-Related Diseases Vs. Evidence Types",
       x = "Disease",
       y = "Number of Associations",
       fill = "Evidence Type") +
  scale_fill_brewer(palette = "Accent")+
  theme_light()

# Basically different types of Lymphoma are clinical+variant associated, and carcinoma is variant associated.
```

```{r}
ggplot(cancer_disease_evidence_counts, aes(x = evidence, y = n, fill = evidence)) +
  geom_bar(stat = "identity") +
  labs(title = "Cancer-Related Diseases Vs. Evidence Types",
       x = "Disease",
       y = "Number of Associations",
       fill = "Evidence Type") +
  scale_fill_brewer(palette = "Accent")+
  theme_light()

# Basically different types of Lymphoma are clinical+variant associated, and carcinoma is variant associated.
```
```{r}
variant_disease <- relations %>%
  filter(entity1_type == "Variant", entity2_type == "Disease")
```

```{r}
total_rows <- nrow(relations)
  unique_genes <- relations %>%
    filter(entity1_type == "Gene") %>%
    distinct(entity1_name) %>%
    nrow()
  unique_variants <- relations %>%
    filter(entity1_type == "Variant") %>%
    distinct(entity1_name) %>%
    nrow()
  unique_diseases <- relations %>%
    filter(entity2_type == "Disease") %>%
    distinct(entity2_name) %>%
    nrow()
```


```{r}
ui <- dashboardPage(
  skin = "green",
  dashboardHeader(title = "Combined Analysis"),
  dashboardSidebar(
    sidebarMenu(
      menuItem("PharmGKB Analysis", tabName = "pharm", icon = icon("chart-pie")),
      menuItem("Summary Stats", tabName = "stats", icon = icon("info-circle")),
      menuItem("Cancer Analysis", tabName = "cancer", icon = icon("chart-bar")),
      menuItem("Network Graph", tabName = "network", icon = icon("project-diagram"))
    )
  ),
  dashboardBody(
    tabItems(
      # Tab 1: PharmGKB Analysis
      tabItem(tabName = "pharm",
              fluidRow(
                box(
                  title = "Comparison Plot Options", 
                  width = 3,
                  selectInput("x1", "Select Entity 1", 
                              choices = c("Gene", "Haplotype", "Variant", "Disease", "Chemical"), 
                              selected = "Gene"),
                  selectInput("y1", "Select Entity 2", 
                              choices = c("Gene", "Haplotype", "Variant", "Disease", "Chemical"), 
                              selected = "Disease")
                ),
                box(
                  title = "Percentage of Association Types Plot", 
                  width = 6,
                  plotOutput("plot1", width = "500px", height = "400px")
                )
              )
      ),
      
      tabItem(tabName = "stats",
              fluidRow(
                infoBoxOutput("totalRowsBox"),
                infoBoxOutput("uniqueGenesBox"),
                infoBoxOutput("uniqueVariantsBox"),
                infoBoxOutput("uniqueDiseasesBox")
                  )
              ),

      
      # Tab 2: Cancer Analysis
      tabItem(tabName = "cancer",
              fluidRow(
                box(
                  title = "Choose Dataset for Cancer Analysis", 
                  width = 3,
                  radioButtons("x2", "Select Dataset", 
                               choices = c("cancer_disease_evidence_counts_v_rs", 
                                           "cancer_disease_evidence_counts_v"), 
                               selected = "cancer_disease_evidence_counts_v_rs")
                ),
                box(
                  title = "Cancer-Related Diseases vs. Evidence Types", 
                  width = 6,
                  plotOutput("plot2", width = "500px", height = "400px")
                )
              )
      ),
      # Tab 3: Network Graph 
      tabItem(tabName = "network",
              fluidRow(
                # Radio buttons to choose ABCB1 or rs1045642
                box(
                  title = "Network Graph Options",
                  width = 3,
                  radioButtons("chosen_entity", "Show network for:",
                               choices = c("ABCB1", "rs1045642"),
                               inline = FALSE)
                ),
                # The network visualization
                box(
                  title = "Interactive Network Graph",
                  width = 9,
                  visNetworkOutput("networkGraph", height = "600px")
                )
              )
      )
    )
  )
)

server <- function(input, output, session) {
  
  # Pie Chart
  output$plot1 <- renderPlot({
    pa <- relations %>% 
      filter(entity1_type == input$x1, entity2_type == input$y1) %>% 
      group_by(association) %>%
      summarise(count = n(), .groups = 'drop') %>%
      mutate(percentage = (count / sum(count)) * 100)
    
    ggplot(pa, aes(x = "", y = percentage, fill = association)) +
      geom_bar(stat = "identity", width = 1) +
      coord_polar("y", start = 0) +
      labs(title = paste(input$x1, "to", input$y1, "Associations"),
           fill = "Association Type") +
      scale_fill_brewer(palette = "Accent") +
      theme_void() +
      geom_text(aes(label = paste0(round(percentage, 1), "%")),
                position = position_stack(vjust = 0.5),
                color = "white", size = 4)
  })
  
  output$totalRowsBox <- renderInfoBox({
    infoBox(
      title = "Total Rows",
      value = total_rows,
      icon = icon("database"),
      color = "blue"
    )
  })

  output$uniqueGenesBox <- renderInfoBox({
    infoBox(
      title = "Unique Genes",
      value = unique_genes,
      icon = icon("dna"),
      color = "purple"
    )
  })

  output$uniqueVariantsBox <- renderInfoBox({
    infoBox(
      title = "Unique Variants",
      value = unique_variants,
      icon = icon("code-branch"),
      color = "fuchsia"
    )
  })

  output$uniqueDiseasesBox <- renderInfoBox({
    infoBox(
      title = "Unique Diseases",
      value = unique_diseases,
      icon = icon("heartbeat"),
      color = "red"
    )
  })
  
  #Cancer Analysis Plot
  output$plot2 <- renderPlot({
    data_to_plot <- switch(input$x2,
                           "cancer_disease_evidence_counts_v_rs" = cancer_disease_evidence_counts_v_rs,
                           "cancer_disease_evidence_counts_v" = cancer_disease_evidence_counts_v)
    
    ggplot(data_to_plot, aes(x = evidence, y = n, fill = evidence)) +
      geom_bar(stat = "identity") +
      labs(title = "Cancer-Related Diseases vs. Evidence Types",
           x = "Evidence", y = "Number of Associations",
           fill = "Evidence Type") +
      scale_fill_brewer(palette = "Accent") +
      theme_light()
  })
  
  
  output$networkGraph <- renderVisNetwork({
    
    
    netData <- relations %>%
      filter(
        entity1_type %in% c("Gene", "Variant"),
        entity2_type == "Disease",
        association == "associated",
        entity1_name == input$chosen_entity
      )
    
    
    nodes <- data.frame(
      id = c(input$chosen_entity, unique(netData$entity2_name)),
      label = c(input$chosen_entity, unique(netData$entity2_name)),
      stringsAsFactors = FALSE
    )
    
    
    nodes$label <- sapply(nodes$label, function(x) {
      paste(strwrap(x, width = 15), collapse = "\n")
    })
    
    
    edges <- netData %>%
      select(from = entity1_name, to = entity2_name)
    
    
    visNetwork(nodes, edges) %>%
      visNodes(
        shape = "box",
        labelHighlightBold = TRUE,
        font = list(size = 14, face = "arial", multi = "md")
      ) %>%
      visEdges(arrows = "to", smooth = TRUE) %>%
      visLayout(randomSeed = 123, improvedLayout = TRUE) %>%
      visPhysics(stabilization = TRUE) %>%
      visOptions(highlightNearest = TRUE, nodesIdSelection = TRUE)
  })
  
  
  session$onSessionEnded(stopApp)
}


shinyApp(ui, server)
```

